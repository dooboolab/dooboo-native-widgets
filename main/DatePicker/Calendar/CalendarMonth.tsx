import React, { PropsWithChildren, ReactElement } from 'react';
import { View, ViewStyle } from 'react-native';

const convertDateString = (date: Date): string => {
  const dateString = new Date(date.getTime() - date.getTimezoneOffset() * 60000)
    .toISOString()
    .split('T')[0];
  return dateString;
};

interface Props<T> {
  monthDate: Date; // first Date of month
  style?: ViewStyle;
  renderDay: ({
    date,
    dailyData,
    isCurrentMonth,
    isToday,
    style,
  }: {
    date: Date;
    dailyData?: T;
    isCurrentMonth: boolean;
    isToday: boolean;
    style?: ViewStyle;
  }) => React.ReactElement;
  calendarWidth: number;
  today: Date; // init Date
  dailyCalData?: { [key: string]: T }; // key format : YYYY-MM-DD
}

/**
 * Month component
 */
function CalendarMonth<T>(props: PropsWithChildren<Props<T>>): ReactElement {
  const thisYear = props.monthDate.getFullYear(); // year number of this month's calendar
  const thisMonth = props.monthDate.getMonth(); // month number of this month's calendar
  const startDate = new Date(
    props.monthDate.getFullYear(),
    props.monthDate.getMonth(),
    1,
  );
  startDate.setDate(startDate.getDate() - startDate.getDay());
  const sYear = startDate.getFullYear();
  const sMonth = startDate.getMonth();
  const sDate = startDate.getDate();

  const datesOfMonth: Date[] = []; // Dates to be displayed on this month's calendar

  for (let i = 0; i < 42; i++) {
    const date = new Date(sYear, sMonth, sDate + i);
    datesOfMonth.push(date);
  }

  // render a week
  const renderWeek = (dates: Date[], key: number): React.ReactElement => {
    const week = dates.map((date) => {
      const isCurMonth =
        date.getFullYear() === thisYear && date.getMonth() === thisMonth;
      return (
        <View key={date.getDate()}>
          {props.renderDay({
            date: date,
            dailyData:
              props.dailyCalData &&
              props.dailyCalData[`${convertDateString(date)}`],
            isCurrentMonth: isCurMonth,
            isToday: convertDateString(props.today) === convertDateString(date),
            style: {
              width: props.calendarWidth / 7,
              height: props.calendarWidth / 7,
            },
          })}
        </View>
      );
    });

    return (
      <View
        key={key}
        style={{
          flexDirection: 'row',
          width: props.calendarWidth,
          justifyContent: 'space-between',
          paddingHorizontal: 0,
        }}>
        {week}
      </View>
    );
  };

  const weeks: React.ReactElement[] = [];

  for (let week = 0; week < 6; week++) {
    const week = renderWeek(datesOfMonth.splice(0, 7), weeks.length);
    weeks.push(week);
  }

  return (
    <>
      <View
        style={{
          ...props.style,
        }}>
        {weeks}
      </View>
    </>
  );
}

export default CalendarMonth;
