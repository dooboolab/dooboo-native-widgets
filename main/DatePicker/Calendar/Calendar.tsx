import { FlatList, Text, View, ViewStyle } from 'react-native';

import CalendarMonth from './CalendarMonth';
import CalendarWeekDays from './CalendarWeekDays';
import React from 'react';

const convertDateString = (date: Date): string => {
  const dateString = new Date(date.getTime() - date.getTimezoneOffset() * 60000)
    .toISOString()
    .split('T')[0];
  return dateString;
};

const getMonthList = (
  initMonthDate: Date,
  pastRange = 36,
  futureRange = 36,
): Date[] => {
  const tmpRows: Date[] = [];
  const year = initMonthDate.getFullYear();
  const month = initMonthDate.getMonth();
  for (let i = -pastRange; i <= futureRange; i++) {
    const monthDate = new Date(year, month + i, 1);
    tmpRows.push(monthDate);
  }
  return tmpRows;
};

interface Props {
  onChangeMonth?: (month: Date) => void; // called when changed month
  initDate?: Date; // initial date, if undefined, it's today
  containerStyle?: ViewStyle;
  renderDay: ({
    date,
    isCurrentMonth,
    isToday,
    style,
  }: {
    date: Date;
    isCurrentMonth: boolean;
    isToday: boolean;
    style?: ViewStyle;
  }) => React.ReactElement;
  pastRange?: number; // Number of past months displayed on the calendar
  futureRange?: number; // Number of future months displayed on the calendar
  calendarWidth: number; // calendar width
  titleContent?: (monthFirst: Date) => React.ReactElement;
  weekdayFormat?: 'narrow' | 'short';
}

/**
 * Pure Calendar
 * - using FlatList
 * - infinite months
 * - render a month view in renderItem of Flatlist
 * - renderDay(function) render a day of month
 */
function Calendar(props: Props): React.ReactElement {
  // init date for calendar
  const { initDate = new Date(), pastRange = 36, futureRange = 36 } = props;

  // init month list (data for FlatList)
  const [monthList] = React.useState<Date[]>(
    getMonthList(initDate, pastRange, futureRange),
  );
  const [curMonthFirst, setCurMonthFirst] = React.useState<Date>(initDate);

  // getItemLayout
  const getItemLayout = React.useCallback((data, index) => {
    return {
      length: props.calendarWidth,
      offset: props.calendarWidth * index,
      index,
    };
  }, []);

  const MemoizedCalendarMonth = React.memo(CalendarMonth, () => {
    return true;
  });

  // render a month
  const renderMonthCalendar = React.useCallback(
    ({ item }): React.ReactElement => {
      return (
        <MemoizedCalendarMonth
          monthDate={item}
          renderDay={props.renderDay}
          calendarWidth={props.calendarWidth}
          style={{
            width: props.calendarWidth,
          }}
          today={initDate}
        />
      );
    },
    [],
  );

  const onViewableItemsChanged = React.useCallback(({ viewableItems }) => {
    if (viewableItems.length === 0) return; // 없는 경우 있음. 그냥 무시
    const { item } = viewableItems[0];
    props.onChangeMonth && props.onChangeMonth(item);
    setCurMonthFirst(item);
  }, []);

  const renderYearMonth = React.useCallback(
    (monthFirst): React.ReactElement => {
      if (props.titleContent) {
        return props.titleContent(monthFirst);
      } else {
        return <Text>{`${convertDateString(monthFirst)}`}</Text>;
      }
    },
    [],
  );

  return (
    <View
      style={{
        ...props.containerStyle,
      }}>
      {/** Year & Month */}
      <View>{renderYearMonth(curMonthFirst)}</View>
      {/** Days of Calendar */}
      <CalendarWeekDays
        calendarWidth={props.calendarWidth}
        style={{
          width: props.calendarWidth,
          height: 22,
        }}
        weekdayFormat={props.weekdayFormat}
      />
      <FlatList
        style={{
          width: props.calendarWidth,
        }}
        data={monthList}
        removeClippedSubviews={true}
        initialNumToRender={1}
        windowSize={2}
        horizontal={true}
        pagingEnabled={true}
        onViewableItemsChanged={onViewableItemsChanged}
        viewabilityConfig={{ viewAreaCoveragePercentThreshold: 50 }}
        renderItem={renderMonthCalendar}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
        keyExtractor={(item): string => {
          return `${item.getFullYear()} + ${item.getMonth()}`;
        }}
        initialScrollIndex={pastRange}
        getItemLayout={getItemLayout}
      />
    </View>
  );
}

export default Calendar;
