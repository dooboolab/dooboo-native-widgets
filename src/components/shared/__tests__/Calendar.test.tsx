import * as React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import Calendar from '../CalendarCarousel';

const component = (props?: any) => {
  return (
    <Calendar
      {...props}
    />
  );
};

describe('[Calendar]', () => {
  it('should render without crashing', () => {
    const rendered = renderer.create(component());
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });
});
