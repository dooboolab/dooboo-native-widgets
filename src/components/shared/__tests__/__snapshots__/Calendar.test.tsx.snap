// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Calendar] should render without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text>
    Calendar
  </Text>
</View>
`;
